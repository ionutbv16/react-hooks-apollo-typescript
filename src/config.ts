
export const REACT_APP_GRAPHQL_URL: string = "https://swapi.graph.cool/";
export const NUMBER_GAME_CARDS: number = 2;
export const PARAM_PEOPLE: string = "People";
export const PARAM_STARSHIPS: string = "Starships" ;