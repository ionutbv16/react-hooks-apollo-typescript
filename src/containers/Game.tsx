import React, { useState, useEffect } from 'react';
import { useQuery, useMutation } from '@apollo/react-hooks';
import { withRouter, RouteComponentProps, useParams } from 'react-router-dom';
import gql from 'graphql-tag';
import styled from 'styled-components';
import CardComp from '../components/Card';
import Loading from '../components/Loading';
import { ICards, ParamTypes, SetIsWinner } from '../types';
import { NUMBER_GAME_CARDS, PARAM_PEOPLE } from '../config';

const Wrapper = styled.div` 
  padding: 20px;
`;

const ActionBar = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding-bottom: 30px;
`;

const CardsWrapper = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;

`;

const MUTATION_ADD_ITEM_TO_HISTORY = gql`
  mutation ($game: Any!) {
    addItemToHistory(game: $game) @client
  }
`
 
export const PEOPLE_QUERY = gql`
  query { allPersons {
      id,
      name,
      height
    }
  }
`;

export const STARSHIPS_QUERY = gql`
  query { allStarships {
      id,
      name,
      hyperdriveRating
    }
  }
`;

export const getGameWinner:any = (cards:any, param:any) => {
  let compareColumn: string = 'hyperdriveRating'
  let biggestSoFar: number = 0;
  let setIsWinner: SetIsWinner = { index: 0, winner: false}
  let cardsWinner: ICards[] = [] 
  if (param === PARAM_PEOPLE) {
    compareColumn = 'height'
  }
  cards.forEach ( (card:any, index:number) => {
    if (card[compareColumn] === biggestSoFar) {
          setIsWinner.winner = false	
    }
    if (card[compareColumn] > biggestSoFar) {
      biggestSoFar = card[compareColumn]
      setIsWinner.winner = true	
      setIsWinner.index = index	
    }
  })

  cards.forEach ( (card:any, index:number) => {
    if (index === setIsWinner.index && setIsWinner.winner === true ) {
      cardsWinner.push({...card, type:param, winner: setIsWinner.winner})
    }
    else {
      cardsWinner.push({...card, type:param, winner: false})
    }
  })
  return cardsWinner;

}

const Game: React.FC<RouteComponentProps> = () => {
  let params = useParams<ParamTypes>();
  const [cards, setCards] = useState<ICards[]>([]);

  let executeQuery: any

  if (params.type === PARAM_PEOPLE) {
    executeQuery = PEOPLE_QUERY
  }
  else {
    executeQuery = STARSHIPS_QUERY
  }
  const [ addItemToHistory ] = useMutation(
    MUTATION_ADD_ITEM_TO_HISTORY,
    { variables: { game: cards } }
  )

  const { data, loading } = useQuery<any>(executeQuery);

  useEffect(() => {
    const dbEndpoint: string = (params.type === PARAM_PEOPLE) ? 'Persons' : 'Starships'
    if (data[`all${dbEndpoint}`] && cards.length < 1)  {
      const twoCards: any = data[`all${dbEndpoint}`].sort(() => Math.random() - Math.random()).slice(0, NUMBER_GAME_CARDS)
      setCards( getGameWinner (twoCards, params.type));
      
    }
    if (cards.length > 1) {
      addItemToHistory()
    }
    
  }, [data, cards]);

  const renderCards = () => {
    if (cards.length > 0) {
       return cards.map((item: ICards) => <CardComp key={item.id} id={item.id} name={item.name} height={item.height} hyperdriverating={item.hyperdriveRating} winner={item.winner} type={item.type} />);
    }
  };

  return (
    <Wrapper>    
      <ActionBar>A Winner needs to have unique biggest value, duplicated biggest value makes no winner. </ActionBar>
      <CardsWrapper>
      {loading ? <Loading /> : renderCards()}
      </CardsWrapper>
    </Wrapper>
  );
};

export default Game;
