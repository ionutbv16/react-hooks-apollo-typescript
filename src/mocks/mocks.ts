
import {STARSHIPS_QUERY} from '../containers/Game'; 

export const itemsMock = [
  {"id":"cj0nwtqvrq53k0114b291ao52","name":"Naboo Royal Starship","hyperdriveRating":1.8},
  {"id":"cj0nwtqpzq4tt01142nh7e9i4","name":"Millennium Falcon","hyperdriveRating":0.5},
]
export const mocks = [
  {
    request: {
      query: STARSHIPS_QUERY,
    },
    result: {
      data: {
        allPersons:  itemsMock,
      },
    },
  },
];

export const gameInput = [{"id":"cj0nv9pi6ewi201308xa3cuha","name":"Arvel Crynyd","height":null,"__typename":"Person"},{"id":"cj0nv9pyeewx30130y8lhjw4e","name":"Sly Moore","height":178,"__typename":"Person"}]


export const gameResult =[{"id":"cj0nv9pg3ewho0130tijy6bk9","name":"Boba Fett","height":183,"__typename":"Person","type":"People","winner":true},{"id":"cj0nv9pyeewx30130y8lhjw4e","name":"Sly Moore","height":178,"__typename":"Person","type":"People","winner":false}]