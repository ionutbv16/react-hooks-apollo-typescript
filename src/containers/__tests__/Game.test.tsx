import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/react-testing';
import { MemoryRouter } from 'react-router';
import wait from 'waait';
import Game, { getGameWinner, STARSHIPS_QUERY} from '../Game'; 
import  {itemsMock, mocks, gameInput, gameResult} from '../../mocks/mocks'; 
import { PARAM_PEOPLE } from '../../config';

describe('Test Game Page', () => {
    it('should render Game without error', async () => {
        await act(async () => {
            const component = renderer.create(
            <MockedProvider mocks={mocks} addTypename={false}>
                <MemoryRouter>
                <Game />
                </MemoryRouter>
            </MockedProvider>,
            );
            await wait(0);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    it('function getGameWinner should return proper number of results', () => {
        const input = gameInput
        const param = PARAM_PEOPLE
        const result = gameResult
        expect(getGameWinner(input)).toHaveLength(2);
    });
})
