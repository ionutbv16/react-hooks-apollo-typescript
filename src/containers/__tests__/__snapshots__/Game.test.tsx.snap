// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Game Page should render Game without error 1`] = `
<div
  className="sc-bZQynM biHpAG"
>
  <div
    className="sc-gzVnrw fAPwhn"
  >
    A Winner needs to have unique biggest value, duplicated biggest value makes no winner. 
  </div>
  <div
    className="sc-htoDjs cdkdqL"
  >
    <div
      className="sc-EHOje dyPYBQ"
    >
      <div
        className="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render Game without error 1`] = `
<div
  className="sc-bZQynM biHpAG"
>
  <div
    className="sc-gzVnrw fAPwhn"
  >
    A Winner needs to have unique biggest value, duplicated biggest value makes no winner. 
  </div>
  <div
    className="sc-htoDjs cdkdqL"
  >
    <div
      className="sc-EHOje dyPYBQ"
    >
      <div
        className="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
